### ISSUE TITLE
Generate documentation from gluon modules

### ISSUE THREAD

**Issue by Marwes** [gluon-lang/gluon#387]
Most of the building blocks should already be in place as the repl and language server uses the same parts to retrieve types and documentation. The way it is structured it can be rather difficult to extract the information into a form amenable for documentation generation though.

## Comment retrieval

Documentation comments are stored in the AST during parsing and can be retrieved with the [metadata](https://docs.rs/gluon_check/0.6.2/gluon_check/metadata/fn.metadata.html) function. So to just get these comments it is enough to parse with [parse_expr](https://docs.rs/gluon/0.6.2/gluon/struct.Compiler.html#method.parse_expr) and then retrieve using `metadata`. The returned structure can then be iterated over and the comments can be printed out in whatever the desired format is.

## Type retrieval

For types we can just call [env_type_of](https://docs.rs/gluon_base/0.6.2/gluon_base/ast/trait.Typed.html#tymethod.env_type_of) to get the type of the expression (make sure it has been run through [typechecking](https://docs.rs/gluon/0.6.2/gluon/struct.Compiler.html#method.typecheck_str). Then, asuming that your module returns a record, you can just iterate (recursively) over the fields of the record type via [row_iter](https://docs.rs/gluon_base/0.6.2/gluon_base/types/struct.ArcType.html#method.row_iter) and [type_field_iter](https://docs.rs/gluon_base/0.6.2/gluon_base/types/struct.ArcType.html#method.type_field_iter).

Side note: [all_symbols](https://docs.rs/gluon_completion/0.6.2/gluon_completion/fn.all_symbols.html) *almost* does what is needed but it returns the results without any qualified names so there is no way to associate the bindings to the `metadata`.

```
// all_symbols would return the bindings `x` but metadata would locate the binding as `inner.x`
let x = Int
{ inner = { x } }
```

[find_info](https://github.com/gluon-lang/gluon/blob/ddd25ae3f249c7a81024d69ce4f652b859f20914/repl/src/repl.rs#L57-L94) which is part of the repl does retrieve the  types that we want

TODO

More?

**Comment by endeav0r** [comment]
A code example of fetching and printing this information?

**Comment by Marwes** [comment]
Something like this should be done for each module you want to generate docs for.

```rust
let (typ, expr) = Compiler::new()
    .typecheck_str(&vm,  module_name, module, None)
    .unwrap();
let meta  = metadata(&*vm.get_env(), &expr);

fn print(typ: &ArcType, meta: &Metadata) {
for field in typ.row_iter() {
    let field_name: &str = field.name.as_ref();
    let meta: &Metadata = meta.module.get(field_name).unwrap();
    let field_type = &field.typ;

    // print whatever

    // Print for submodules
    print(field.typ, meta);
}
}
```

**Comment by endeav0r** [comment]
Thanks!

**Comment by Marwes** [comment]
Closed by #456 (and some other PRs)
